src/shell.rs

```diff
@@ -36,219 +91,205 @@ pub enum ColorChoice {
 }
 
 // Should simplify the APIs a lot.
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct MessageInfo {
     pub color_choice: ColorChoice,
     pub verbosity: Verbosity,
+    pub stdout_needs_erase: bool,
+    pub stderr_needs_erase: bool,
 }
 
 impl MessageInfo {
+    pub const fn new(color_choice: ColorChoice, verbosity: Verbosity) -> MessageInfo {
+        MessageInfo {
+            color_choice,
+            verbosity,
+            stdout_needs_erase: false,
+            stderr_needs_erase: false,
+        }
+    }
+
     pub fn create(verbose: bool, quiet: bool, color: Option<&str>) -> Result<MessageInfo> {
         let color_choice = get_color_choice(color)?;
         let verbosity = get_verbosity(color_choice, verbose, quiet)?;
 
         Ok(MessageInfo {
             color_choice,
             verbosity,
+            stdout_needs_erase: false,
+            stderr_needs_erase: false,
         })
     }
 
-    pub fn verbose(self) -> bool {
+    pub fn is_verbose(&self) -> bool {
         self.verbosity.verbose()
     }
-}
 
-impl Default for MessageInfo {
-    fn default() -> MessageInfo {
-        MessageInfo {
-            color_choice: ColorChoice::Auto,
-            verbosity: Verbosity::Normal,
-        }
-    }
-}
+    fn as_verbosity<T, C: Fn(&mut MessageInfo) -> T>(&mut self, call: C, new: Verbosity) -> T {
```

Comment:
I've added these functions so we can temporarily change the verbosity of a function call, either to silence or to be more verbose, etc. It allows us to store the old value, and return a result and ensure the old state is restored.